### ISSUE TITLE
Can you provide some details about these lines in yolov5_tflite

### ISSUE THREAD

**Issue by masouduut94** [zldrobit/onnx_tflite_yolov3#13]
Hello sir

I am working on an object detection project and currently have come across your code. but I don't understand these lines from [this one](https://github.com/zldrobit/yolov5/blob/4f7551fce4f48f0f8685a5358fd484a250acf2fa/detect.py#L232) to [this one](https://github.com/zldrobit/yolov5/blob/4f7551fce4f48f0f8685a5358fd484a250acf2fa/detect.py#L248).

```
grid = [torch.zeros(1)] * nl  # init grid
a = torch.tensor(anchors).float().view(nl, -1, 2).to(device)
anchor_grid = a.clone().view(nl, 1, -1, 1, 2)  # shape(nl,1,na,1,2)
z = []  # inference output
for i in range(nl):
    _, _, ny_nx, _ = x[i].shape
    r = imgsz[0] / imgsz[1]
    nx = int(np.sqrt(ny_nx / r))
    ny = int(r * nx)
    grid[i] = _make_grid(nx, ny).to(x[i].device)
    stride = imgsz[0] // ny
    y = x[i].sigmoid()
    y[..., 0:2] = (y[..., 0:2] * 2. - 0.5 + grid[i].to(x[i].device)) * stride  # xy
    y[..., 2:4] = (y[..., 2:4] * 2) ** 2 * anchor_grid[i]  # wh
    z.append(y.view(-1, no))

```
I could not open an issue in the same yolov5 that you have forked from the original repo, so I opened this one in here.

**Comment by zldrobit** [comment]
These lines of code is basically building output from previous convolutions, which is part of the yolov5 `Detect` module.
```
a = torch.tensor(anchors).float().view(nl, -1, 2).to(device)
anchor_grid = a.clone().view(nl, 1, -1, 1, 2)  # shape(nl,1,na,1,2)
```
builds anchors, which read from yaml files (I will change it to read from *.pt weights https://github.com/ultralytics/yolov5/pull/1127#issuecomment-714671350).

```
    _, _, ny_nx, _ = x[i].shape
    r = imgsz[0] / imgsz[1]
    nx = int(np.sqrt(ny_nx / r))
    ny = int(r * nx)
```
gets the output shape ny (vertical) and nx (horizontal) of the *i*-th yolo head.

```
    grid[i] = _make_grid(nx, ny).to(x[i].device)
```
creates [meshgrid](https://www.tensorflow.org/api_docs/python/tf/meshgrid) from the shape ny and nx.

```
    y[..., 0:2] = (y[..., 0:2] * 2. - 0.5 + grid[i].to(x[i].device)) * stride  # xy
    y[..., 2:4] = (y[..., 2:4] * 2) ** 2 * anchor_grid[i]  # wh
```
are the equations to output center coordinates xy and, width and height of the output bounding boxes.

Or you could post your question on the PRs:
https://github.com/ultralytics/yolov5/pull/959
https://github.com/ultralytics/yolov5/pull/1127

**Comment by masouduut94** [comment]
thanks for the quick reply.
The truth is that I am stuck with a dilemma that I can't figure out a solution to.
I trained a yolov5 model for mask detection service. In the inference step I get bounding boxes like this one:

![Capture](https://user-images.githubusercontent.com/21217928/97835278-0c0ef380-1cef-11eb-9e54-dd8f25aa8806.PNG)

You noticed that the height is too big.

And the whole code goes here:

```
import numpy as np
import tensorflow as tf

from utils import common_utils as cu
from utils.datasets import letterbox


class YoloV5_tflite:

    def __init__(self, config):
        self.anchors = self.get_anchors()
        self.model_path = config['model_path']
        self.input_shape = config['input_shape']
        self.target_shape = config['target_shape']
        self.conf_thresh = config['conf_thresh']
        self.iou_thresh = config['iou_thresh']
        self.classes_len = config['classes_len']
        self.anchor_len = config['anchor_len']
        self.interpreter, self.input_details, self.output_details = \
            cu.load_tflite_model(self.model_path)

        gpus = tf.config.experimental.list_physical_devices('GPU')
        print(gpus)
        if gpus:
            # Restrict TensorFlow to only use the first GPU
            try:
                tf.config.experimental.set_visible_devices(gpus[0], 'GPU')
                logical_gpus = tf.config.experimental.list_logical_devices('GPU')
                print(len(gpus), "Physical GPUs,", len(logical_gpus), "Logical GPU")
            except RuntimeError as e:
                # Visible devices must be set before GPUs have been initialized
                print(e)

    def inference(self, image):
        self.target_shape = image.shape[:2]

        img_input = self.resize_img(image)

        self.input_shape = img_input.shape[1:3]

        self.interpreter.set_tensor(self.input_details[0]['index'], img_input)
        self.interpreter.invoke()

        x = [self.interpreter.get_tensor(self.output_details[i]['index'])
             for i in range(1, self.anchor_len + 1)]

        y = self.get_pred(x, [self.input_shape[0], self.input_shape[1]])
        y = self.non_max_suppression(y)
        boxes, scores, classes = [], [], []
        if y[0] is not None:
            y = np.squeeze(y, axis=0)
            boxes, scores, classes = y[:, :4], y[:, 4:5], y[:, 5:6]
            boxes = self.scale_boxes(boxes)

        return boxes, scores, classes

    def scale_boxes(self, boxes):
        gain = min(self.input_shape[0] / self.target_shape[0],
                   self.input_shape[1] / self.target_shape[1])
        pad = (self.input_shape[1] - self.target_shape[1] * gain) / 2, \
              (self.input_shape[0] - self.target_shape[0] * gain) / 2
        boxes[:, [0, 2]] -= pad[0]
        boxes[:, [1, 3]] -= pad[1]
        boxes[:, :] /= gain

        return boxes

    def outputs_to_numpy(self, boxes, scores, classes):

        boxes = boxes.detach().cpu().numpy()
        scores = scores.detach().cpu().numpy()
        classes = classes.detach().cpu().numpy()
        return boxes, scores, classes

    def resize_img(self, image):
        """
        Resize input image to the expected input shape
        """
        image_resized = letterbox(image, new_shape=self.input_shape, auto=False)[0]
        # image_resized = tf.image.resize(image,
        #                                 [self.input_shape[0], self.input_shape[1]])
        image_np = image_resized / 255.0
        image_exp = np.expand_dims(image_np, axis=0)
        image_exp = tf.cast(image_exp, dtype=tf.float32)

        return image_exp

    def make_grid(self, nx, ny):
        xv, yv = np.meshgrid(np.arange(nx), np.arange(ny))
        return np.stack((xv, yv), axis=2).reshape((1, 1, ny * nx, 2))

    def get_anchors(self):
        anchors = [[10, 13, 16, 30, 33, 23],
                   [30, 61, 62, 45, 59, 119],
                   [116, 90, 156, 198, 373, 326]]
        a = np.array(anchors).reshape(3, -1, 2)
        anchor_grid = np.copy(a).reshape((3, 1, -1, 1, 2))  # .astype(np.float32)

        return anchor_grid

    def get_pred(self, x, image_size):
        z = []
        grid = [np.zeros(1)] * self.anchor_len

        for i in range(self.anchor_len):
            _, _, ny_nx, _ = x[i].shape
            ratio = image_size[1] / image_size[0]
            # ratio = image_size[0] / image_size[1]
            ny = int(np.sqrt(ny_nx / ratio))
            nx = int(ratio * ny)  # + 1
            grid[i] = self.make_grid(nx, ny)
            stride = image_size[0] // ny
            y = tf.math.sigmoid(x[i]).numpy()
            y[..., 0:2] = (y[..., 0:2] * 2. - 0.5 + grid[i]) * stride  # xy
            y[..., 2:4] = (y[..., 2:4] * 2) ** 2 * self.anchors[i]
            z.append(y.reshape((-1, 7)))

        pred = np.concatenate((z[0], z[1], z[2]), axis=0)
        pred = np.expand_dims(pred, axis=0)
        return pred

    def scale_coords(self, img1_shape, coords, img0_shape, ratio_pad=None):
        # Rescale coords (xyxy) from img1_shape to img0_shape
        if ratio_pad is None:  # calculate from img0_shape
            gain = min(img1_shape[0] / img0_shape[0], img1_shape[1] / img0_shape[1])  # gain  = old / new
            pad = (img1_shape[1] - img0_shape[1] * gain) / 2, (img1_shape[0] - img0_shape[0] * gain) / 2  # wh padding
        else:
            gain = ratio_pad[0][0]
            pad = ratio_pad[1]

        coords[:, [0, 2]] -= pad[0]  # x padding
        coords[:, [1, 3]] -= pad[1]  # y padding
        coords[:, :4] /= gain
        return coords

    def decode_box(self, x):
        # Convert nx4 boxes from [x, y, w, h] to [x1, y1, x2, y2] where xy1=top-left, xy2=bottom-right
        y = np.zeros_like(x)  # if isinstance(x, torch.Tensor) else np.zeros_like(x)
        y[:, 0] = x[:, 0] - x[:, 2] / 2  # top left x
        y[:, 1] = x[:, 1] - x[:, 3] / 2  # top left y
        y[:, 2] = x[:, 0] + x[:, 2] / 2  # bottom right x
        y[:, 3] = x[:, 1] + x[:, 3] / 2  # bottom right y
        return y

    def non_max_suppression(self, prediction):
        """Performs Non-Maximum Suppression (NMS) on inference results

        Returns:
             detections with shape: nx6 (x1, y1, x2, y2, conf, cls)
        """
        xc = prediction[..., 4] > self.conf_thresh  # candidates

        # Settings
        max_wh = 4096  # (pixels) minimum and maximum box width and height

        output = [None]
        for xi, x in enumerate(prediction):  # image index, image inference
            # Apply constraints
            x = x[xc[xi]]  # confidence

            # If none remain process next image
            if not x.shape[0]:
                continue

            # Compute conf
            x[:, 5:] *= x[:, 4:5]  # conf = obj_conf * cls_conf

            # Box (center x, center y, width, height) to (x1, y1, x2, y2)
            box = self.decode_box(x[:, :4])

            i, j = np.nonzero(x[:, 5:] > self.conf_thresh)  # .nonzero(as_tuple=False).T
            i, j = np.transpose(i), np.transpose(j)
            # i, j = (x[:, 5:] > self.conf_thresh).nonzero(as_tuple=False).T
            x = np.concatenate((box[i], x[i, j + 5, None], j[:, None]), axis=1)
            # Batched NMS
            c = x[:, 5:6] * max_wh  # classes
            boxes, scores = x[:, :4] + c, x[:, 4]  # boxes (offset by class), scores
            keep = cu.single_class_non_max_suppression(boxes, scores, self.conf_thresh, self.iou_thresh)
            if len(keep) > 0:
                output[xi] = x[keep]

        return np.asarray(output)

    def filter_boxes(self, idxs, boxes):
        """
        Filter out the boxes selected by the NMS, by their index.

        Args:
            idxs (list): List of indexes
            boxes (list): List of bounding boxes

        Returns:
            (list): List of filtered boxes
        """
        filter_boxes = []
        for i in idxs:
            box = self.scale_boxes(boxes[i])
            filter_boxes.append(box)

        return filter_boxes

```

and `commonutils.non_maximum_suppression` is in here:

```
def single_class_non_max_suppression(
        bboxes,
        confidences,
        conf_thresh=0.2,
        iou_thresh=0.5,
        keep_top_k=-1):
    """
    Do Non Max Suppresssion on single class.
    For the specific class, given the bbox and its confidence,

    1) sort the bbox according to the confidence from top to down, we call this
    a set

    2) select the bbox with the highest confidence, remove it from set, and do
    IOU calculate with the rest bbox

    3) remove the bbox whose IOU(Intersection of Union) is higher than
        the iou_thresh from the set,

    4) loop step 2 and 3, until the set is empty.

    Leaves only the bounding box with the highest confidence.

    Args:
        bboxes (np.array): bboxes numpy array of 2D, [num_bboxes, 4]
        confidences (np.array): Confidences numpy array of 1D. [num_bboxes]
        conf_thresh (float): Confidence threshold (0.2 by default)
        iou_thresh (float): Intersection of Union threshold (0.5 by default)
        keep_top_k (int): How many bboxes to keep, in case of -1 keeps all
            of them (-1 by default)

    Returns:
        (list): List of indexes which should be kept
    """
    if len(bboxes) == 0:
        return []

    conf_keep_idx = np.where(confidences > conf_thresh)[0]

    bboxes = bboxes[conf_keep_idx]
    confidences = confidences[conf_keep_idx]

    # pick bboxes with highest confidence
    pick_highest_conf = []
    xmin = bboxes[:, 0]
    ymin = bboxes[:, 1]
    xmax = bboxes[:, 2]
    ymax = bboxes[:, 3]

    area = (xmax - xmin + 1e-3) * (ymax - ymin + 1e-3)
    idxs = np.argsort(confidences)

    # pick bboxes with highest confidence,
    # remove bboxes with IOU higher than threshold,
    # so we are picking bbox with highest confidence
    # for each
    while len(idxs) > 0:
        last = len(idxs) - 1
        i = idxs[last]
        pick_highest_conf.append(i)

        # keep top k
        if keep_top_k != -1:
            if len(pick_highest_conf) >= keep_top_k:
                break

        overlap_xmin = np.maximum(xmin[i], xmin[idxs[:last]])
        overlap_ymin = np.maximum(ymin[i], ymin[idxs[:last]])
        overlap_xmax = np.minimum(xmax[i], xmax[idxs[:last]])
        overlap_ymax = np.minimum(ymax[i], ymax[idxs[:last]])
        overlap_w = np.maximum(0, overlap_xmax - overlap_xmin)
        overlap_h = np.maximum(0, overlap_ymax - overlap_ymin)
        overlap_area = overlap_w * overlap_h

        overlap_ratio = overlap_area / (area[idxs[:last]] + area[i] - overlap_area)

        need_to_be_deleted_idx = np.concatenate(
            ([last], np.where(overlap_ratio > iou_thresh)[0]))

        idxs = np.delete(idxs, need_to_be_deleted_idx)

    # if the number of final bboxes is less than keep_top_k, we need to pad it.
    # TODO
    return conf_keep_idx[pick_highest_conf]
```

I know that's a little bit tedius to look at others code, but currently, I am desperate to find solution. It would be so kind of you to guide me to overcome this.

**Comment by zldrobit** [comment]
Close this issue due to there're other suitable places for this issue (https://github.com/ultralytics/yolov5/pull/959, https://github.com/ultralytics/yolov5/pull/1127).
